### PR TITLE
Add "Gateway Injection" feature

### PR DESCRIPTION
**Pull request by howardjohn**

### CODE REVIEW

File: features/gateway_injection.md [hunk 1]
```diff
@@ -0,0 +1,146 @@
+# Feature: Gateway Injection
```
Comment (brian-avery) [inline features/gateway_injection.md:1]: Can you please add this to [features.yaml](https://github.com/istio/enhancements/blob/master/features.yaml)?

Comment (brian-avery) [inline features/gateway_injection.md:1]: My intention was for the 

# Feature to just be a heading, but everyone is adding the feature name after it. I think it might make sense to change the template and remove `**Feature name:**`

File: features/gateway_injection.md [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+### Requirements:
+
+- [ ] [User stories](insert_your_link_here) reviewed in a work group meeting.
+
+- [x] [RFC Authored] - [Injector V2](https://docs.google.com/document/d/1Rmp9B3DDypgMCau-YAqidx_r3qvKLZj6jUX-t22zj84/).
+
+- [x] [Documentation](https://preliminary.istio.io/latest/docs/setup/additional-setup/sidecar-injection/#custom-templates-experimental) and [here](https://preliminary.istio.io/latest/docs/setup/additional-setup/external-controlplane/) for enabling and using the feature.
+
+- [ ] [Feedback plan](insert_your_link_here). ??
```
Comment (brian-avery) [inline features/gateway_injection.md:21]: This one is really only useful for Experimental. Designed so that when we create experimental features, we have an idea of how we will tell if this feature is useful for users and if anything needs to change before it goes alpha.

**Comment by brian-avery** [comment]
@googlebot refresh

**Comment by brian-avery** [comment]
@howardjohn how's this looking? I think all this needs is the update to features.yaml?

**Review by brian-avery — APPROVED** [review]

File: features/gateway_injection.md [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+## Experimental
+
+### Requirements:
+
+- [ ] [User stories](insert_your_link_here) reviewed in a work group meeting.
+
+- [x] [RFC Authored] - [Injector V2](https://docs.google.com/document/d/1Rmp9B3DDypgMCau-YAqidx_r3qvKLZj6jUX-t22zj84/).
+
+- [x] [Documentation](https://preliminary.istio.io/latest/docs/setup/additional-setup/sidecar-injection/#custom-templates-experimental) and [here](https://preliminary.istio.io/latest/docs/setup/additional-setup/external-controlplane/) for enabling and using the feature.
```
Comment (smawson) [inline features/gateway_injection.md:19]: Couple comments:
1) First doc is about custom sidecar templates, it does not mention using this for gateways or provide an example. So I think it doesn't really qualify as documentation for this feature.
2) Second doc is about external control planes, is this only usable with external control planes? That doesn't sound right.

File: features.yaml [hunk 4]
```diff
@@ -29,6 +29,12 @@ features:
       maturity: Stable
       nextExpectedPromotion: ""
     area: Traffic Management
+  - name: "Gateway Injection"
+    level:
+      checklist: ""
+      maturity: Stable
```
Comment (smawson) [inline features.yaml:35]: Fix maturity. I assume this is either alpha or experimenta?

File: features/gateway_injection.md [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+- [x] Explicit user action is required to enable this feature (e.g. a config field, config resource, or installation action).
+
+> [Documentation](https://preliminary.istio.io/latest/docs/setup/additional-setup/sidecar-injection/#custom-templates-experimental)
+
+**Docs**
+
+- [x] Reference docs are published to preliminary.istio.io or the Istio wiki.
+  
+There are no reference docs required for this feature.
```
Comment (smawson) [inline features/gateway_injection.md:47]: I think there is a reference doc for the creation of custom templates you could link to. Also link to reference doc for the templates label.

File: features/gateway_injection.md [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+**Short description:** Injecting gateway deployments to enabled minimal configuration and simplified operations.
+
+---
+
+## Experimental
+
+### Requirements:
+
+- [ ] [User stories](insert_your_link_here) reviewed in a work group meeting.
```
Comment (smawson) [inline features/gateway_injection.md:15]: User stories are in the RFC I think, could just link to that.

File: features/gateway_injection.md [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+- [x] Design doc describing the intention of the feature, how it will be
+  implemented, and any thoughts on how to test the feature has been approved by
+  relevant work group leads. [Design Doc](https://docs.google.com/document/d/1Rmp9B3DDypgMCau-YAqidx_r3qvKLZj6jUX-t22zj84/).
+- [ ] Feature coverage and test plans written and approved.
+
+**Docs**
+
+- [ ] Documentation on istio.io includes performance expectations; may have caveats.
+- [x] Documentation on istio.io includes samples/tutorials. [Here](https://preliminary.istio.io/latest/docs/setup/additional-setup/sidecar-injection/#custom-templates-experimental) and [here](https://preliminary.istio.io/latest/docs/setup/additional-setup/external-controlplane/) for enabling and using the feature.
+- [x] Documentation on istio.io includes appropriate glossary entries. No glossary entries needed, this uses existing concepts.
```
Comment (smawson) [inline features/gateway_injection.md:88]: There are no glossary entries for gateway or injection, I think we should probably add some simple ones for those. So I would un-check-mark this since the existing concepts don't have entries either :)

File: features/gateway_injection.md (hunk 3 above)
Comment (howardjohn) [reply features/gateway_injection.md:19]: https://github.com/istio/istio.io/pull/9706 is specifically for this, I will update this PR with a link to that for docs once its merged.

For 2, it is already the documented way to use gateways external control plane, but can be used in general. I think the above PR will clarify this a lot hopefully

File: features.yaml (hunk 4 above)
Comment (howardjohn) [reply features.yaml:35]: Oops, thanks. It want to go to alpha ideally, but can put experimental if we are not ready. will update

Comment (frankbu) [reply features.yaml:35]: Since external control plane is alpha and is using it, it would be best to make this alpha as well. Otherwise, wouldn't we need to demote external control plane?

Comment (smawson) [reply features.yaml:35]: Let's try for alpha, put it on agenda for next TOC to get the TOC approval.

Comment (nrjpoddar) [reply features.yaml:35]: What's the level of testing we have done for this feature @howardjohn ?

Injection templates have been flaky and used for several things, I think we should start with experimental here unless we have higher confidence of this working.

File: features/gateway_injection.md [hunk 8]
```diff
@@ -0,0 +1,146 @@
+# Feature
+
+**Feature name:** Gateway Injection
+
+**Primary lead(s):** @howardjohn
+
+**Short description:** Injecting gateway deployments to enabled minimal configuration and simplified operations.
```
Comment (nrjpoddar) [inline features/gateway_injection.md:7]:
```suggestion
**Short description:** Injecting gateway deployments to enable minimal configuration and simplified operations.
```

**Review by nrjpoddar — COMMENTED** [review]
I saw the link to test and it feels quite minimal given the scope of changes specially around customization and revision based deployments. I would suggest adding more tests before marking this alpha.

File: features.yaml (hunk 4 above)
Comment (howardjohn) [reply features.yaml:35]: We have e2e tests that ensure we can inject a gateway through various installation methods (helm and plain yaml) and that traffic works, etc. We also have extensive unit tests of injection.

I am not sure what you mean by injection templates have been flaky - they are one of the oldest features of Istio.

Additionally, this is the only mechanism to install gateways for external control plane, so we have a high level of testing of this feature by end users.

The alpha bar:
"The feature is covered by unit tests and integration tests where the feature is enabled and the tests are non-flaky. Tests may not cover all corner cases, but the most common cases have been covered." which I feel we have met.

File: features/gateway_injection.md (hunk 7 above)
Comment (howardjohn) [reply features/gateway_injection.md:88]: Wow, very surprised - I assumed they did. Opened https://github.com/istio/istio.io/pull/9733

**Comment by howardjohn** [comment]
@smawson @nrjpoddar I updated with new references as the docs are merged. I think that was the primary outstanding concern but let me know if anything else is needed

**Review by nrjpoddar — APPROVED** [review]
